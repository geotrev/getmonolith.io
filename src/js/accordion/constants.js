export const Selectors = {
  // unique
  DATA_ACCORDION: "data-accordion",
  DATA_COLLAPSIBLE: "data-collapsible",
  DATA_TOGGLE_MULTIPLE: "data-toggle-multiple",
  // common
  DATA_PARENT: "data-parent",
  DATA_VISIBLE: "data-visible",
  DATA_TARGET: "data-target",
  // accessibility
  ARIA_EXPANDED: "aria-expanded",
  ARIA_HIDDEN: "aria-hidden",
}

export const Messages = {
  NO_ACCORDION_ID_ERROR:
    "Could not initialize accordion; you must include a value for the 'data-accordion' attribute.",
  NO_ACCORDION_ERROR: id => `Could not find element matching [data-accordion='${id}']`,
  TRIGGERS_TO_COLLAPSIBLES_LENGTH_WARNING: (id, triggersLength, collapsiblesLength) =>
    `Your accordion with id '${id}' has ${triggersLength} triggers and ${collapsiblesLength} collapsibles; make sure there is a trigger for each collapsible!`,
}
