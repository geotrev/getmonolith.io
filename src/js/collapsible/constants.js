export const Selectors = {
  // unique
  DATA_COLLAPSIBLE: "data-collapsible",
  // common
  DATA_VISIBLE: "data-visible",
  DATA_TARGET: "data-target",
  DATA_PARENT: "data-parent",
  // accessibility
  ARIA_EXPANDED: "aria-expanded",
  ARIA_CONTROLS: "aria-controls",
  ARIA_HIDDEN: "aria-hidden",
  ARIA_LABELLEDBY: "aria-labelledby",
  TABINDEX: "tabindex",
  // classes
  TRIGGER_READY_CLASS: "collapsible-trigger-ready",
  CONTENT_READY_CLASS: "collapsible-content-ready",
}

export const CssProperties = {
  HEIGHT: "height",
  VISIBILITY: "visibility",
}

export const CssValues = {
  AUTO: "auto",
  HIDDEN: "hidden",
  VISIBLE: "visible",
}

export const Events = {
  CLICK: "click",
  TRANSITIONEND: "transitionend",
}

export const Messages = {
  NO_COLLAPSIBLE_ID_ERROR:
    "Could not initialize collapsible; you must include a value for the 'data-collapsible' attribute.",
  NO_TRIGGER_ERROR: id =>
    `Could not find collapsible trigger with [data-target='${id}']; you can't have a collapsible without a trigger.`,
  NO_TRIGGER_ID_ERROR: id => `Could not find id on collapsible trigger with [data-target='${id}'].`,
  NO_CONTENT_ERROR: id =>
    `Could not find collapsible content with id '${id}'; you can't have a collapsible without content.`,
}
