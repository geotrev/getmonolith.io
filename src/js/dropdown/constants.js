export const KeyCodes = {
  TAB: 9,
  SHIFT: 16,
  ESCAPE: 27,
  ARROW_UP: 38,
  ARROW_DOWN: 40,
}

export const Selectors = {
  // unique
  DATA_DROPDOWN: "data-dropdown",
  DROPDOWN_MENU_CLASS: "dropdown-menu",
  // common
  DATA_TARGET: "data-target",
  DATA_PARENT: "data-parent",
  DATA_VISIBLE: "data-visible",
  // accessibility
  TABINDEX: "tabindex",
  ARIA_HASPOPUP: "aria-haspopup",
  ARIA_CONTROLS: "aria-controls",
  ARIA_LABELLEDBY: "aria-labelledby",
  ARIA_EXPANDED: "aria-expanded",
  // classes
  OVERLAY_OPEN: "overlay-open",
}

export const Events = {
  KEYDOWN: "keydown",
  CLICK: "click",
  BLUR: "blur",
}

export const Messages = {
  NO_DROPDOWN_ID_ERROR:
    "Could not setup dropdown. Make sure it has a valid [data-dropdown] attribute with a unique id as its value.",
  NO_MENU_ERROR: attr => `Could not find menu associated with ${attr}.`,
  NO_DROPDOWN_ITEMS_ERROR: attr => `Could not find any list items associated with ${attr}.`,
  NO_DROPDOWN_ACTIONS_ERROR: attr =>
    `Could not find any button or anchor elements associated with ${attr}.`,
  NO_PARENT_ERROR: "Could not find dropdown button's [data-parent] attribute.",
}
