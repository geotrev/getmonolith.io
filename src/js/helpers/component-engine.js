import { queryAll } from "./dom"
import { isBrowserEnv } from "./is-browser-env"

const isString = value => typeof value === "string"
const isFunction = value => typeof value === "function"

export const ComponentEngine = {
  /**
   * Initialize an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   */
  start: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg } = metadata

    if (id && isString(id)) {
      const instance = document.querySelector(`[${attribute}='${id}']`)
      if (!instance) return

      const validComponent = [instance].filter(thisArg._validate)[0]
      if (!validComponent) return

      thisArg._components.push(validComponent)
    } else if (!id && !thisArg._components.length) {
      const instances = queryAll(`[${attribute}]`)
      if (!instances.length) return

      const validComponents = instances.filter(thisArg._validate)
      thisArg._components = thisArg._components.concat(validComponents)
    } else {
      // attempted to .start() when .stop() wasn't run,
      // OR tried to instantiate a component that's already active.
    }
  },
  /**
   * Teardown an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   * @property {string} activeNodeKey
   * @property {*=} cancelActiveFn
   */
  stop: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg, activeNodeKey, cancelActiveFn } = metadata

    if (id && isString(id)) {
      let targetIndex
      const instance = thisArg._components.filter((activeInstance, index) => {
        if (activeInstance.getAttribute(attribute) !== id) return false
        targetIndex = index
        return true
      })[0]

      if (!instance) return

      if (
        thisArg[activeNodeKey] &&
        instance === thisArg[activeNodeKey] &&
        isFunction(cancelActiveFn)
      )
        thisArg[cancelActiveFn]()

      thisArg._teardown(instance)
      thisArg._components.splice(targetIndex, 1)
    } else if (!id && thisArg._components.length) {
      if (thisArg[activeNodeKey] && isFunction(cancelActiveFn)) thisArg[cancelActiveFn]()

      thisArg._components.forEach(thisArg._teardown)
      thisArg._components = []
    }
  },
}
