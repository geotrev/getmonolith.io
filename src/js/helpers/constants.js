export const KeyCodes = {
  SHIFT: 16,
  TAB: 9,
  ARROW_UP: 38,
  ARROW_DOWN: 40,
}

export const Selectors = {
  FOCUSABLE_TAGS: ["a", "button", "input", "object", "select", "textarea", "[tabindex]"],
  KEYBOARD_CLASS: "using-keyboard",
  NOT_VISUALLY_HIDDEN_CLASS: ":not(.is-visually-hidden)",
  TABINDEX: "tabindex",
}

export const Events = {
  KEYDOWN: "keydown",
  CLICK: "click",
  BLUR: "blur",
}

export const Messages = {
  NO_SELECTOR_STRING_OR_CHILDREN_ERROR:
    "createFocusTrap must be given one or both of: first parameter (as selector string)" +
    " and/or options.children (array of elements).",
  OPTION_MATCHERS_DATA_TYPE_ERROR:
    "Invalid data type given to options.matchers for createFocusTrap. Expected: Array.",
  INCORRECT_MATCHER_TYPE_ERROR: type =>
    `Invalid matcher given to options.matchers for createFocusTrap. Expected: String. Recieved: ${type}.`,
  NO_MATCHER_LENGTH_ERROR:
    "Invalid value given to options.matchers for createFocusTrap; value must be an array with at least one selector string",
}
