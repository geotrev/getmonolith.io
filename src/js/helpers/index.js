export * from "./component-engine"
export { createFocusRing } from "./create-focus-ring"
export { createFocusTrap } from "./create-focus-trap"
export * from "./dom"
export { focusOnce } from "./focus-once"
export { isBrowserEnv } from "./is-browser-env"
export { isiOSMobile } from "./is-ios-mobile"
export { log } from "./log"
export { throttle } from "./throttle"
