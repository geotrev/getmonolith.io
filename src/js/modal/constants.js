export const KeyCodes = {
  ESCAPE: 27,
}

export const Selectors = {
  // unique
  DATA_MODAL: "data-modal",
  // common
  DATA_TARGET: "data-target",
  DATA_VISIBLE: "data-visible",
  DATA_CLOSE: "data-close",
  DATA_PARENT: "data-parent",
  // accessibility
  ARIA_HIDDEN: "aria-hidden",
  ARIA_MODAL: "aria-modal",
  ROLE: "role",
  TABINDEX: "tabindex",
  // classes
  NO_SCROLL: "no-scroll",
}

export const CssProperties = {
  PADDING_RIGHT: "paddingRight",
  PADDING_LEFT: "paddingLeft",
  CURSOR: "cursor",
}

export const CssValues = {
  AUTO: "auto",
  POINTER: "pointer",
}

export const Events = {
  KEYDOWN: "keydown",
  CLICK: "click",
  RESIZE: "resize",
  BLUR: "blur",
}

export const Messages = {
  NO_TRIGGER_ERROR: id => `Could not find modal trigger with id ${id}.`,
  NO_ID_ERROR:
    "Could not detect an id on your [data-modal] element. Please add a value matching the modal trigger's [data-parent] attribute.",
  NO_MODAL_DIALOG_ERROR: id => `Could not find element with attribute [data-parent='${id}'].`,
}
