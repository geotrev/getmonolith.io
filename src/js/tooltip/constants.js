export const KeyCodes = {
  ESCAPE: 27,
}

export const Selectors = {
  // unique
  DATA_TOOLTIP: "data-tooltip",
  // common
  DATA_VISIBLE: "data-visible",
  DATA_TARGET: "data-target",
  // accessibility
  ROLE: "role",
  ARIA_DESCRIBEDBY: "aria-describedby",
  // classes
  DROP_INLINE_START_CLASS: "is-drop-inline-start",
  DROP_INLINE_END_CLASS: "is-drop-inline-end",
  OVERLAY_OPEN: "overlay-open",
}

export const CssProperties = {
  HEIGHT: "height",
  WIDTH: "width",
  TOP: "top",
  LEFT: "left",
}

export const Events = {
  CLICK: "click",
  MOUSEOVER: "mouseover",
  MOUSEOUT: "mouseout",
  FOCUS: "focus",
  BLUR: "blur",
  KEYDOWN: "keydown",
}

export const Messages = {
  NO_ID_ERROR: "Could not find tooltip id.",
  NO_TRIGGER_ERROR: id => `Could not find a tooltip trigger with attribute [data-target='${id}'].`,
  NO_TOOLTIP_ERROR: id => `Could not find a tooltip with id '${id}'.`,
}
