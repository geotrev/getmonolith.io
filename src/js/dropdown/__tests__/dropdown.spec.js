import Undernet from "../../"
import { renderDOM, simulateKeyboardEvent } from "../../helpers/test"
import { KeyCodes, Messages } from "../constants"

const dom = `
  <div data-dropdown="dropdown-1" class="dropdown">
    <button id="trigger-id-1" data-parent="dropdown-1" data-target="dropdown-menu-1">Open Dropdown</button>
    <ul id="dropdown-menu-1" class="dropdown-menu">
      <li><a href="#">Item 1</a></li>
      <li><a href="#">Item 2</a></li>
      <li><a href="#">Item 3</a></li>
    </ul>
  </div>
`

console.error = jest.fn()

describe("Dropdown", () => {
  describe("API start", () => {
    it("sets attributes", () => {
      // Given
      const wrapper = renderDOM(dom)
      // When
      Undernet.Dropdowns.start()
      // Then
      expect(wrapper).toMatchSnapshot()
      Undernet.Dropdowns.stop()
    })
  })

  describe("API stop + #handleClick", () => {
    it("does not open dropdown", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = document.querySelector("#trigger-id-1")
      // When
      Undernet.Dropdowns.start()
      Undernet.Dropdowns.stop()
      trigger.click()
      // Then
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe("#handleClick", () => {
    let wrapper, firstItem

    beforeEach(() => {
      wrapper = renderDOM(dom)

      const trigger = document.querySelector("#trigger-id-1")
      firstItem = document.querySelector("#dropdown-menu-1 a")

      Undernet.Dropdowns.start()
      trigger.click()
    })

    afterEach(() => Undernet.Dropdowns.stop())

    it("opens clicked dropdown", () => {
      expect(wrapper).toMatchSnapshot()
    })

    it("sets focus to the first dropdown link", () => {
      expect(document.activeElement).toEqual(firstItem)
    })
  })

  describe("#handleArrowKeyPress", () => {
    let wrapper, trigger, firstItem, lastItem

    beforeEach(() => {
      wrapper = renderDOM(dom)

      trigger = document.querySelector("#trigger-id-1")
      const items = document.querySelectorAll("#dropdown-menu-1 a")
      firstItem = items[0]
      lastItem = items[items.length - 1]

      Undernet.Dropdowns.start()
    })

    afterEach(() => Undernet.Dropdowns.stop())

    it("opens dropdown on down arrow key press", () => {
      // When
      simulateKeyboardEvent(KeyCodes.ARROW_DOWN, false, trigger)
      // Then
      expect(wrapper).toMatchSnapshot()
    })

    it("opens dropdown on up arrow key press", () => {
      // When
      simulateKeyboardEvent(KeyCodes.ARROW_UP, false, trigger)
      // Then
      expect(wrapper).toMatchSnapshot()
    })

    it("sets focus to first item in dropdown menu on down arrow key press", () => {
      // When
      simulateKeyboardEvent(KeyCodes.ARROW_DOWN, false, trigger)
      // Then
      expect(document.activeElement).toEqual(firstItem)
    })

    it("sets focus to first item in dropdown menu on up arrow key press", () => {
      // When
      simulateKeyboardEvent(KeyCodes.ARROW_UP, false, trigger)
      // Then
      expect(document.activeElement).toEqual(lastItem)
    })
  })

  describe("#handleFirstTabClose & #handleLastTabClose", () => {
    let wrapper, firstItem, lastItem

    beforeEach(() => {
      wrapper = renderDOM(dom)

      const trigger = document.querySelector("#trigger-id-1")
      const items = document.querySelectorAll("#dropdown-menu-1 a")
      firstItem = items[0]
      lastItem = items[items.length - 1]

      Undernet.Dropdowns.start()
      trigger.click()
    })

    afterEach(() => Undernet.Dropdowns.stop())

    it("closes dropdown if shift + tab key is pressed in open menu while first child is focused", () => {
      // When
      simulateKeyboardEvent(KeyCodes.TAB, true, firstItem)
      // Then
      expect(wrapper).toMatchSnapshot()
    })

    it("closes dropdown if tab key is pressed in open menu while last child is focused", () => {
      // When
      simulateKeyboardEvent(KeyCodes.TAB, false, lastItem)
      // Then
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe("#handleClose", () => {
    let wrapper
    let trigger

    beforeEach(() => {
      wrapper = renderDOM(dom)

      trigger = document.querySelector("#trigger-id-1")

      Undernet.Dropdowns.start()
      trigger.click()
      document.querySelector("#dropdown-menu-1 a").click()
    })

    afterEach(() => Undernet.Dropdowns.stop())

    it("closes dropdown", () => {
      expect(wrapper).toMatchSnapshot()
    })

    it("sets focus back to dropdown button", () => {
      expect(document.activeElement).toEqual(trigger)
    })
  })

  describe("#handleEscapeKeyPress", () => {
    it("closes dropdown", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = document.querySelector("#trigger-id-1")
      // When
      Undernet.Dropdowns.start()
      trigger.click()
      simulateKeyboardEvent(KeyCodes.ESCAPE)
      // Then
      expect(wrapper).toMatchSnapshot()
      Undernet.Dropdowns.stop()
    })
  })

  describe("#handleOffMenuClick", () => {
    it("closes dropdown", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = document.querySelector("#trigger-id-1")
      // When
      Undernet.Dropdowns.start()
      trigger.click()
      document.querySelector("body").click()
      // Then
      expect(wrapper).toMatchSnapshot()
      Undernet.Dropdowns.stop()
    })
  })

  describe("#handleReturnFocus", () => {
    it("sets focus to dropdown trigger", () => {
      // Given
      renderDOM(dom)
      const trigger = document.querySelector("#trigger-id-1")
      // When
      Undernet.Dropdowns.start()
      trigger.click()
      simulateKeyboardEvent(KeyCodes.ESCAPE)
      // Then
      expect(document.activeElement).toEqual(trigger)
      Undernet.Dropdowns.stop()
    })
  })

  describe("Multiple Dropdowns", () => {
    let wrapper

    beforeEach(() => {
      const newDOM = dom.concat(`
        <div data-dropdown="dropdown-2" class="dropdown">
          <button id="trigger-id-2" data-parent="dropdown-2" data-target="dropdown-menu-2">Open Dropdown 2</button>
          <ul id="dropdown-menu-2" class="dropdown-menu">
            <li><a href="#">Item 1</a></li>
          </ul>
        </div>
      `)
      wrapper = renderDOM(newDOM)

      const trigger1 = document.querySelector("#trigger-id-1")
      const trigger2 = document.querySelector("#trigger-id-2")

      Undernet.Dropdowns.start()
      trigger1.click()
      trigger2.click()
    })

    afterEach(() => Undernet.Dropdowns.stop())

    it("closes first dropdown if second dropdown is clicked", () => {
      expect(wrapper).toMatchSnapshot()
    })

    it("sets focus to first dropdown item of second dropdown", () => {
      const firstItem = document.querySelector("#dropdown-menu-2 a")
      expect(document.activeElement).toEqual(firstItem)
    })
  })
})

// prettier-ignore
const errorDom = (
  dropdown,
  buttonId,
  parent,
  target,
  menuId,
  hasUl = true,
  hasItems = true,
  hasButtons = true
) => `
  <div data-dropdown="${dropdown}" class="dropdown">
    <button id="${buttonId}" data-parent="${parent}" data-target="${target}">Open Dropdown 2</button>
    ${hasUl && `<ul id="${menuId}" class="dropdown-menu">
      ${hasItems && `<li>
        ${hasButtons && `<a href="#">Item 1</a>` || ""}
      </li>` || ""}
    </ul>` || ""}
  </div>
`

describe("Dropdown Warnings", () => {
  const DROPDOWN_ATTR = '[data-dropdown="dropdown-id"]'

  it("prints console error if dropdown id can't be found", () => {
    // Given
    renderDOM(errorDom("", "trigger-id", "dropdown-id", "menu-id", "menu-id"))
    // When
    Undernet.Dropdowns.start()
    // Then
    expect(console.error).toBeCalledWith(Messages.NO_DROPDOWN_ID_ERROR)
  })

  it("prints console error if dropdown menu can't be found", () => {
    // Given
    renderDOM(errorDom("dropdown-id", "trigger-id", "dropdown-id", "menu-id", "menu-id", false))
    // When
    Undernet.Dropdowns.start()
    // Then
    expect(console.error).toBeCalledWith(Messages.NO_MENU_ERROR(DROPDOWN_ATTR))
  })

  it("prints console error if dropdown items can't be found", () => {
    // Given
    renderDOM(
      errorDom("dropdown-id", "trigger-id", "dropdown-id", "menu-id", "menu-id", true, false)
    )
    // When
    Undernet.Dropdowns.start()
    // Then
    expect(console.error).toBeCalledWith(Messages.NO_DROPDOWN_ITEMS_ERROR(DROPDOWN_ATTR))
  })

  it("prints console error if dropdown buttons or links can't be found", () => {
    // Given
    renderDOM(
      errorDom("dropdown-id", "trigger-id", "dropdown-id", "menu-id", "menu-id", true, true, false)
    )
    // When
    Undernet.Dropdowns.start()
    // Then
    expect(console.error).toBeCalledWith(Messages.NO_DROPDOWN_ACTIONS_ERROR(DROPDOWN_ATTR))
  })

  it("prints console error if dropdown trigger's [data-target] attribute cant' be found", () => {
    // Given
    renderDOM(errorDom("dropdown-id", "trigger-id", "", "menu-id", "menu-id"))
    // When
    Undernet.Dropdowns.start()
    // Then
    expect(console.error).toBeCalledWith(Messages.NO_PARENT_ERROR)
  })
})
