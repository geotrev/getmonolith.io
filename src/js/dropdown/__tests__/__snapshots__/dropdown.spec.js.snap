// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown #handleArrowKeyPress opens dropdown on down arrow key press 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="true"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="true"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleArrowKeyPress opens dropdown on up arrow key press 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="true"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="true"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleClick opens clicked dropdown 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="true"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="true"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleClose closes dropdown 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
      tabindex="-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleEscapeKeyPress closes dropdown 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-id-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="new-dropdown"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-id-1"
      data-target="new-dropdown"
      id="trigger-id-1"
      tabindex="-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="new-dropdown"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown2"
  >
    
    
    <button
      aria-controls="dropdown-id-2"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown2"
      data-target="dropdown-id-2"
      id="trigger-id-2"
    >
      Open Dropdown 2
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-2"
      class="dropdown-menu"
      id="dropdown-id-2"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleFirstTabClose & #handleLastTabClose closes dropdown if shift + tab key is pressed in open menu while first child is focused 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
      tabindex="-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleFirstTabClose & #handleLastTabClose closes dropdown if tab key is pressed in open menu while last child is focused 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
      tabindex="-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown #handleOffMenuClick closes dropdown 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-id-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="new-dropdown"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-id-1"
      data-target="new-dropdown"
      id="trigger-id-1"
      tabindex="-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="new-dropdown"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown2"
  >
    
    
    <button
      aria-controls="dropdown-id-2"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown2"
      data-target="dropdown-id-2"
      id="trigger-id-2"
    >
      Open Dropdown 2
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-2"
      class="dropdown-menu"
      id="dropdown-id-2"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown API start sets attributes 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown API stop + #handleClick does not open dropdown 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

</body>
`;

exports[`Dropdown Multiple Dropdowns closes first dropdown if second dropdown is clicked 1`] = `
<body>
  
  
  <div
    class="dropdown"
    data-dropdown="dropdown-1"
    data-visible="false"
  >
    
    
    <button
      aria-controls="dropdown-menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      data-parent="dropdown-1"
      data-target="dropdown-menu-1"
      id="trigger-id-1"
    >
      Open Dropdown
    </button>
    
    
    <ul
      aria-labelledby="trigger-id-1"
      class="dropdown-menu"
      id="dropdown-menu-1"
    >
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 1
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 2
        </a>
      </li>
      
      
      <li>
        <a
          href="#"
          tabindex="-1"
        >
          Item 3
        </a>
      </li>
      
    
    </ul>
    
  
  </div>
  

        
  <div
    class="dropdown"
    data-dropdown="dropdown-2"
    data-visible="true"
  >
    
          
    <button
      aria-controls="dropdown-menu-2"
      aria-expanded="true"
      aria-haspopup="true"
      data-parent="dropdown-2"
      data-target="dropdown-menu-2"
      id="trigger-id-2"
    >
      Open Dropdown 2
    </button>
    
          
    <ul
      aria-labelledby="trigger-id-2"
      class="dropdown-menu"
      id="dropdown-menu-2"
    >
      
            
      <li>
        <a
          href="#"
          tabindex="0"
        >
          Item 1
        </a>
      </li>
      
          
    </ul>
    
        
  </div>
  
      
</body>
`;
